scene/resources/surface_tool.cpp

```diff
@@ -623,55 +621,50 @@ void SurfaceTool::index() {
 	}
 
 	HashMap<Vertex, int, VertexHasher> indices;
-	List<Vertex> new_vertices;
+	LocalVector<Vertex> old_vertex_array = vertex_array;
+	vertex_array.clear();
 
-	for (List<Vertex>::Element *E = vertex_array.front(); E; E = E->next()) {
-		int *idxptr = indices.getptr(E->get());
+	for (uint32_t i = 0; i < old_vertex_array.size(); i++) {
+		int *idxptr = indices.getptr(old_vertex_array[i]);
 		int idx;
 		if (!idxptr) {
 			idx = indices.size();
-			new_vertices.push_back(E->get());
-			indices[E->get()] = idx;
+			vertex_array.push_back(old_vertex_array[i]);
+			indices[old_vertex_array[i]] = idx;
 		} else {
 			idx = *idxptr;
 		}
 
 		index_array.push_back(idx);
 	}
 
-	vertex_array.clear();
-	vertex_array = new_vertices;
-
 	format |= Mesh::ARRAY_FORMAT_INDEX;
```

Comment:
Will you be adding a quantized mesh format? You'd have to read up what gltf quantized requires. https://github.com/KhronosGroup/glTF/issues/1670

Reply:
Here are a list of relevant specs from gltf.

* https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression
* https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression
* https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/MSFT_lod

Reply:
Nope, we do our own optimizations on import

Reply:
To be clear, I'm asking about the benefit of smaller in memory sizes for meshes. I don't think it's possible to use a smaller type for vertices.

> As an example, a static PBR-ready mesh typically requires POSITION (12 bytes), TEXCOORD (8 bytes), NORMAL (12 bytes) and TANGENT (16 bytes) for each vertex, for a total of 48 bytes. With this extension, it is possible to use SHORT to store position and texture coordinate data (8 and 4 bytes, respectively) and BYTE to store normal and tangent data (4 bytes each), for a total of 20 bytes per vertex with often negligible quality impact.

Reply:
Godot 4.0 uses:
* 12 bytes for position. At some point we added quantizing for this as optional, but it just generally looks bad and users had a lot of problems. Meshes had visible seams, small details get deformed easily, etc. Its just not worh it.
* 4 bytes for normal (Godot 4.0 uses ARGB2101010)
* 4 bytes for tangent (also ARGB2101010)
* 8 bytes for UV. Again, we s upported having less, but to be honest users nowadays mostly use larger textures than 1024, so using half float for UV looks pretty terrible.

This way, godot uses 12+4+4+8=28 bytes per vertex instead of 20. Its honestly not much of a problem and things just work.